### PR TITLE
API Generator

### PR DESCRIPTION
**Pull request by nhtruong**
### Description
Added API Generator to generate API actions from OpenSearch spec.

### Issues Resolved
closes https://github.com/opensearch-project/opensearch-ruby/issues/139

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: api_generator/lib/api_generator.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+require_relative 'action'
+require_relative 'action_generator'
+require_relative 'spec_generator'
+require_relative 'namespace_generator'
+require_relative 'index_generator'
+
+# Generate API endpoints for OpenSearch Ruby client
+class ApiGenerator
+  HTTP_VERBS = %w[get post put patch delete patch head].freeze
+  EXISTING_NAMESPACES = Set.new(%w[
```
Comment (dblock) [inline api_generator/lib/api_generator.rb:19]: Externalize into a set of folders/files (e.g. load a list of folders checked into git).

**Review by dblock — COMMENTED** [review]
This is very good. I would try harder to externalize anything hard-coded like lists of API methods, and I think we still need:

1. An (easier) way to run the generator via `rake generate`. 
2. A CI/CD that fails when the APIs are out of date.
3. A cron that updates the APIs, ala https://github.com/slack-ruby/slack-ruby-client/blob/master/.github/workflows/update_api.yml

**Comment by nhtruong** [comment]
@dblock Agree with all 3. They are actually in the roadmap. We will have follow up PR to add those features. For now, this is to be done manually so that the Ruby client can get the new endpoints added easily.

File: api_generator/USER_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+---
+
+### Usage
+This generator should be run everytime the OpenSearch API Specification is updated to propagate the changes to the Ruby client. For now, this must be done manually:
+- Create a new branch from `main`
+- Download the latest OpenSearch API Specification from [The API Spec Repo](https://github.com/opensearch-project/opensearch-api-specification/blob/main/OpenSearch.openapi.json)
+- Run the generator with the API Spec downloaded previously (see below)
+- Run Rubocop with `-a` flag to remove redundant spacing from the generated code `rubocop -a`
+- Commit and create a PR to merge the updated API actions into `main`.
```
Comment (dblock) [inline api_generator/USER_GUIDE.md:11]: Nit, copy-paste commands to do this as an example so people can "just do it".

**Review by dblock — APPROVED** [review]
Some nits on docs and an initializer, I skimmed the rest, I think it's good enough and we should move forward.

File: api_generator/lib/action.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+require_relative 'parameter'
+
+# A collection of operations that comprise a single API Action
+class Action
+  attr_reader :group, :name, :namespace, :http_verbs, :urls, :description, :external_docs,
+              :parameters, :path_params, :query_params,
+              :body, :body_description, :body_required
+
+  # @param [Array<Operation>] operations
+  def initialize(operations)
```
[the remark this answers is not in the document]
Comment (nhtruong) [reply api_generator/lib/action.rb:20]: No there are multiple operations in 1 action, but they all share the group property, that's broken into name and namespace

**Comment by nhtruong** [comment]
@dblock Will address these in a follow-up PR. I'm going to switch over getting the security API generated after this is merged.

**Review by VachaShah — APPROVED** [review]
This looks great @nhtruong! 

Can we also list down the next steps/improvements in a new issue or may be in the existing one #139?

**Pull request merged**